File: packages/js/ethereal-genui-core/src/message_segments.ts
/// Splits an assistant message into ordered segments so fenced blocks render as
/// sibling widgets. Streaming-tolerant: a trailing fence that hasn't closed yet
/// yields a segment with closed === false.

export type MessageSegment = TextSegment | CodeSegment | UiSegment

export interface TextSegment {
  readonly kind: 'text'
  readonly markdown: string
}

export interface CodeSegment {
  readonly kind: 'code'
  readonly language: string
  readonly code: string
  readonly closed: boolean
}

export interface UiSegment {
  readonly kind: 'ui'
  readonly json: string
  readonly closed: boolean
}

const _fenceOpen = /^\s*```(.*)$/
const _fenceClose = /^\s*```\s*$/

/// Parse content into text / code / ui segments. Streaming-tolerant: a
/// trailing fence that hasn't closed yet yields a segment with closed === false.
export function parseSegments(content: string): MessageSegment[] {
  const segments: MessageSegment[] = []
  const lines = content.split('\n')
  let textBuffer = ''

  function flushText(): void {
    const text = textBuffer
    textBuffer = ''
    // trimRight equivalent: remove trailing whitespace
    const trimmed = text.replace(/\s+$/, '')
    if (trimmed.trim().length > 0) {
      segments.push({ kind: 'text', markdown: trimmed })
    }
  }

  let i = 0
  while (i < lines.length) {
    const openMatch = _fenceOpen.exec(lines[i])
    if (openMatch !== null) {
      flushText()
      const lang = openMatch[1].trim()
      i++
      const body: string[] = []
      let closed = false
      while (i < lines.length) {
        if (_fenceClose.test(lines[i])) {
          closed = true
          i++
          break
        }
        body.push(lines[i])
        i++
      }
      const inner = body.join('\n')
      if (lang.toLowerCase() === 'ui') {
        segments.push({ kind: 'ui', json: inner, closed })
      } else {
        segments.push({ kind: 'code', language: lang, code: inner, closed })
      }
    } else {
      textBuffer += lines[i] + '\n'
      i++
    }
  }
  flushText()
  return segments
}

File: lib/src/message_segments.dart
/// Splits an assistant message into ordered segments so fenced blocks render as
/// sibling widgets (outside `gpt_markdown`'s `Text.rich`, where taps were eaten).
sealed class MessageSegment {
  const MessageSegment();
}

class TextSegment extends MessageSegment {
  const TextSegment(this.markdown);
  final String markdown;
}

class CodeSegment extends MessageSegment {
  const CodeSegment(this.language, this.code, this.closed);
  final String language;
  final String code;
  final bool closed;
}

class UiSegment extends MessageSegment {
  const UiSegment(this.json, this.closed);
  final String json;
  final bool closed;
}

final _fenceOpen = RegExp(r'^\s*```(.*)$');
final _fenceClose = RegExp(r'^\s*```\s*$');

/// Parse [content] into text / code / ui segments. Streaming-tolerant: a
/// trailing fence that hasn't closed yet yields a segment with `closed == false`.
List<MessageSegment> parseSegments(String content) {
  final segments = <MessageSegment>[];
  final lines = content.split('\n');
  final textBuffer = StringBuffer();

  void flushText() {
    final text = textBuffer.toString();
    textBuffer.clear();
    if (text.trim().isNotEmpty) segments.add(TextSegment(text.trimRight()));
  }

  var i = 0;
  while (i < lines.length) {
    final open = _fenceOpen.firstMatch(lines[i]);
    if (open != null) {
      flushText();
      final lang = open.group(1)!.trim();
      i++;
      final body = <String>[];
      var closed = false;
      while (i < lines.length) {
        if (_fenceClose.hasMatch(lines[i])) {
          closed = true;
          i++;
          break;
        }
        body.add(lines[i]);
        i++;
      }
      final inner = body.join('\n');
      if (lang.toLowerCase() == 'ui') {
        segments.add(UiSegment(inner, closed));
      } else {
        segments.add(CodeSegment(lang, inner, closed));
      }
    } else {
      textBuffer.writeln(lines[i]);
      i++;
    }
  }
  flushText();
  return segments;
}
